### PR TITLE
sign in and sign up

### PR DESCRIPTION
**Pull request by pranavkps**

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    const hashedPassword = await bcryptjs.hash(password,10);
+
+    const result = await userModel.create({
+      email :email,
+      password :hashedPassword,
+      username :username
+    });
+
+    const token = jwt.sign({email :result.email,id : result._id},SECRET_KEY);
+    res.status(201).json({user : result, token :token});
```
Comment (srikanthkatta07) [inline src/controllers/userController.js]: Just return token field only user not required.

File: src/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
+const { connect } = require('mongoose');
+const userModel = require('../models/user.model');
+const jwt = require('jsonwebtoken');
+// const bcrypt = require('bcryptjs');
+const bcryptjs = require('bcryptjs');
+
+const SECRET_KEY = "NOTESAPI";
```
Comment (srikanthkatta07) [inline src/controllers/userController.js]: Move secret key to env file

File: src/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  try{
+    const exsistingUser = await userModel.findOne({email : email});
+    if(!exsistingUser){
+      return res.status(404).json({message : "user not found"});
+    }
+
+    const matchPassword = await bcryptjs.compare(password,exsistingUser.password);
+
+    if(!matchPassword){
+      return res.status(404).json({message : "user not found"});
```
Comment (srikanthkatta07) [inline src/controllers/userController.js]: Give message saying "Given password is wrong."

File: src/controllers/userController.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    const token = jwt.sign({email : exsistingUser.email, id : exsistingUser._id},SECRET_KEY);
+    res.status(201).json({user : exsistingUser, token :token});
```
Comment (srikanthkatta07) [inline src/controllers/userController.js]: Status code should be 200.
Return token only, user not required.

File: src/models/user.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+//     type : String,
+//     required : true
+//   },
+//   email :{
+//     type : String,
+//     required : true
+//   }
+// } ,{timestamps : true});
+
+// module.exports = mongoose.model("User", UserSchema);
```
Comment (srikanthkatta07) [inline src/models/user.js]: Remove this file.

File: src/routes/v1/userRouter.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+const express = require('express');
+const { signin, signup } = require('../../controllers/userController');
+const userRouter = express.Router();
+
+userRouter.post('/signup', signup);
+
+userRouter.post('./signin', signin);
```
Comment (srikanthkatta07) [inline src/routes/v1/userRouter.js]: why './signin'? shouldn't it be '/signin'

File: src/models/user.js (hunk 5 above)
Comment (srikanthkatta07) [reply src/models/user.js]: Are we using existing userModel object. Just have a look as it contains other fields also like role, name etc.

File: src/app.js [hunk 7]
```diff
@@ -59,4 +60,13 @@ app.use(errorConverter);
 // handle error
 app.use(errorHandler);
 
+//user
+app.use('/users',userRouter)
+
+app.get("/api",(req,res)=>{
```
Comment (srikanthkatta07) [inline src/app.js]: Just remove this route "/api" as we are not using this api.

File: src/controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+const signup = async (req,res)=>{
+
+  const {username,email,password} = req.body;
+  try{
+    const exsistingUser = await userModel.findOne({email : email});
+    if(exsistingUser){
+      return res.status(400).json({message : "user already exsist"});
+    }
+
+    const hashedPassword = await bcryptjs.hash(password,10);
```
Comment (maharjunm) [inline src/controllers/userController.js:18]: all the logic here should move to service file.

File: src/controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// sign-in code
+
+const signin = async (req,res)=>{
+
+  const { email , password } = req.body;
+
+  try{
+    const exsistingUser = await userModel.findOne({email : email});
```
Comment (maharjunm) [inline src/controllers/userController.js:43]: here also. logic should move into service file.

File: src/controllers/userController.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
+const { connect } = require('mongoose');
+const userModel = require('../models/user.model');
+const jwt = require('jsonwebtoken');
+// const bcrypt = require('bcryptjs');
```
Comment (maharjunm) [inline src/controllers/userController.js]: remove commented code.

File: src/services/userServices.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  try{
+    const exsistingUser = await userModel.findOne({email : email});
+    if(!exsistingUser){
+      return res.status(404).json({message : "user not found"});
+    }
+
+    const matchPassword = await bcryptjs.compare(password,exsistingUser.password);
+
+    if(!matchPassword){
+      return res.status(404).json({message : "Given Password is Wrong"});
```
Comment (srikanthkatta07) [inline src/services/userServices.js]: Return 400 only

**Review by srikanthkatta07 — APPROVED** [review]

**Pull request merged**